### PR TITLE
Bathtub Overhaul

### PR DESCRIPTION
**Pull request by pgmzeta**
[CLEANLINESS][FEATURE]

## About the PR 

This PR overhauls bathtubs by cleaning up code, updating audio/visual cues, and changing interaction mechanics.

Code cleanup
* Use UpdateOverlays instead of src.overlays
* Spread code into procs
* Removed 'Draw a Bath', 'Eject Occupant' local verbs

Presentation
* Liquid overlay appears above and below mob
* Fluid depth conveyed via Fluid overlay, transparency
* Audio/text cues for most interactions
* Flavor: Bloodies mob clothing in a tub of blood

Mechanics
* Change: Cyborgs cannot pull people out of tubs
* Change: Emptying a bathtub applies a TOUCH effect to those without gloves
* Change: Bathtubs can transfer all fluid like a regent container
* Change: Applies reagents via TOUCH instead of direct transfer
* Feature: The flip emote splashes tub contents on the below turf
* Feature: Interaction now starts/stops a tap that fills up the tub over time
* Feature: Laying down in a tub acts like laying down in the same fluid on a turf
* Flavor: 1% chance of a soda-water bath on creation

Balance
In addition to testing the changes and features, I confirmed bath bombs work, as well as not suffocating when bioeffects, mutantrace, or internals would prevent it. During light testing, no one acid/fire/etc mix I put together was able to outright kill someone in a single process, including vampires with holy water. That said - if you have a tub full of i.e. sarin, you can put a lot of units into a mob in one process by click-dragging them in.

## Why's this needed? 
Bathtub interactions simulate both turf fluid & reagent container in-game mechanics.
Small changes in interactions for quality of life and better emulating bathtubs we're familiar with.
Better visual indication of bathtub fluid levels.
Light flavor to make them more interesting.

## In-game Example
As of commit [a5058df](https://github.com/goonstation/goonstation/pull/7853/commits/a5058dff27274cf39a30ab0414d04a3b36cd0153)
https://www.youtube.com/watch?v=Kkwaj5c-0mE

Update as of [112052f] - tap functionality, audio/visual effects
https://www.youtube.com/watch?v=bxDgcQtRhDM

## Changelog

```changelog
(u)glowbold
(*) Bathtubs now apply reagents to occupants as turf fluid puddles would - but you can only drown while laying.
(+) Bathtubs now have a tap, visually fill up over time, and occupants now appear under the fluid.
```

### CODE REVIEW

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 1]
```diff
@@ -95,16 +95,31 @@
...
+		. = ..()
+
+	mob_flip_inside(var/mob/user)
+		if (src.reagents.total_volume)
+			user.show_text("<span class='alert'>You splash around enough to shake the tub!</span>")
+			src.visible_message("<span class='notice'>[src.occupant] splish-splashes around!</span>")
+			playsound(src.loc, "sound/impact_sounds/Liquid_Slosh_1.ogg", 25, 1)
+			animate_wiggle_then_reset(src, 1, 3)
+			src.reagents.trans_to(src.last_turf, 5)
+			return
```
Comment (ZeWaka) [inline code/WorkInProgress/DrMelonsStuff.dm:120]: usle.
```suggestion
```

File: code/mob/living/life/breath.dm [hunk 2]
```diff
@@ -102,7 +102,11 @@
 			return // no breathing inside possessed objects
 		else if (istype(owner.loc, /obj/machinery/atmospherics/unary/cryo_cell))
 			return
-
+		else if (istype(owner.loc, /obj/machinery/bathtub) && owner.lying)
+			if (owner.loc.reagents.total_volume > 300)
+				var/obj/fluid/F = new
+				F.reagents = owner.loc.reagents
+				underwater = F
```
Comment (frawhst) [inline code/mob/living/life/breath.dm:109]: I was going to review this like 2 days ago but I couldn't think of a clean way to do what you want here.
TL;DR this won't work- fluids track reagents via their `group`, rather than each fluid obj tracking their own reagents. This means that later on, when `force_mob_to_ingest` tries to make the occupant drink from the fluid they're drowning in, it'll see the fluid doesn't have a group and return without doing anything.

I suppose you could override the bathtub's `handle_internal_lifeform` to handle drowning "manually", but that would require some code duplication. grumble

**Comment by frawhst** [comment]
oh, looks good visually, but changelog is pretty log- should trim that down to 1 or 2 entries

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 3]
```diff
@@ -221,7 +274,8 @@
     set src in oview(1)
     set category = "Local"
     if (get_dist(usr, src) <= 1 && !usr.stat)
-        src.reagents.add_reagent(default_reagent,120)
+        playsound(src.loc, "sound/impact_sounds/Liquid_Slosh_2.ogg", 50, 1)
+        src.reagents.add_reagent(default_reagent, 100)
```
[the remark this answers is not in the document]
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: slightly smaller steps give a bit smoother transparency gradient, and also fits evenly into the current max of 500u. 

thank you for asking here, it made me think what that verb was trying to convey - the change I just pushed feels like a cleaner representation.

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 4]
```diff
@@ -114,117 +147,166 @@
...
-
-		var/msg
 
-		if(target == user && !user.stat)	// if drop self, then climbed in
-			msg = "[user.name] climbs into [src]."
-			boutput(user, "<span class='notice'>You climb into [src].</span>")
+		if (!istype(target) || target.buckled || LinkBlocked(target.loc,src.loc) || BOUNDS_DIST(user, src) > 1 || BOUNDS_DIST(user, target) > 1 || is_incapacitated(user) || isAI(user))
+			return
+		if (src.occupant)
+			boutput(user, "<span class='alert'>Someone's already in the [src]!</span>")
```
Comment (ZeWaka) [inline code/WorkInProgress/DrMelonsStuff.dm:295]: i think you accidentally removed a return in here :)

Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm:295]: thank you!!

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 5]
```diff
@@ -114,117 +147,167 @@
...
-			src.overlays += fluid_icon
-
-		if(src.myuser)
-			src.overlays += src.myuser
-			// Don't really see the purpose of this; re-add if you believe otherwise
-			// if(reagents.total_volume)
-				// src.overlays += fluid_icon_feet
-
-		src.overlays += bath_edge
+			var/percent_filled = reagents.total_volume / reagents.maximum_volume
```
Comment (frawhst) [inline code/WorkInProgress/DrMelonsStuff.dm:162]: this isn't really a percent (should be filled_ratio or fraction_filled or something)

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 6]
```diff
@@ -114,117 +147,167 @@
...
+				var/image/new_overlay = src.SafeGetOverlayImage("fluid_overlay", 'icons/obj/stationobjs.dmi', "fluid_bathtub", MOB_LAYER - 0.2)
+				new_overlay.color = average.to_rgba()
+				src.UpdateOverlays(new_overlay, "fluid_overlay")
+
+				// I don't want to do this here but no where else feels right
+				// maybe a new chem transfer instead? Could be a new knob to tweak for cyrox too?
+				// #define SUBMERGED 4
+				// it would also make future projects of mine easier, so both this is and i am a hack
+				if(reagents.get_reagent_amount("blood") > 200) // arbitrary
+					// reagents don't track DNA, so it's your own blood vOv
```
Comment (frawhst) [inline code/WorkInProgress/DrMelonsStuff.dm:180]: Blood does! or it should.
It's stored in the vague `data` var- here's a snippet from stable mutagen (no, I don't know why that var is called `cheating`)
```
		var/datum/reagent/blood/cheating = holder.reagent_list["blood"]
		if (cheating && istype(cheating.data, /datum/bioHolder))
			src.data = cheating.data
```
should be able to grab the blood from the tub via the same method as above, get the bioHolder, and then pass the bioHolder's `owner` to `add_blood` to add the right DNA. Wait, why doesn't `add_blood` accept a bioHolder as an argument? why do we have to do this roundabout get-the-owner-to-get-the-bioHolder thing?

```
// WHAT THE ACTUAL FUCK IS THIS SHIT
// WHO THE FUCK WROTE THIS
/atom/proc/add_blood(atom/source, var/amount = 5)
```
ahhh ok then

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 7]
```diff
@@ -114,117 +147,167 @@
...
-				reagents.reaction(src.myuser, TOUCH)
-				src.reagents.trans_to(src.myuser,5,1)
-			else
-				src.reagents.clear_reagents()
-				on_reagent_change()
-
-		return
+			if(src.reagents.total_volume)
+				src.reagents.reaction(src.occupant, TOUCH)
+				src.reagents.trans_to(src.occupant, 5)
```
Comment (frawhst) [inline code/WorkInProgress/DrMelonsStuff.dm:288]: are we like... absorbing the reagents through our skin? the above line should be enough to get skin-pen chems to add to people, and other chems to do whatever other skin-stuff they do.

Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm:288]: this is preserving the previous behavior (older-diff, line 179) but i am perfectly happy to remove it

File: code/WorkInProgress/DrMelonsStuff.dm (hunk 6 above)
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: :                        )

**Comment by pgmzeta** [comment]
tweaked the changelog again, hopefully this is good :)?

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 8]
```diff
@@ -114,117 +147,178 @@
...
+
+	mouse_drop(atom/over_object as obj)
+		if (istype(over_object, /turf))
+			drain_bathtub(usr)
+			return
+		if (!istype(over_object, /obj/item/reagent_containers/glass) && !istype(over_object, /obj/item/reagent_containers/food/drinks) && !istype(over_object, /obj/reagent_dispensers) && !istype(over_object, /obj/item/spraybottle) && !istype(over_object, /obj/machinery/plantpot) && !istype(over_object, /obj/mopbucket) && !istype(over_object, /obj/item/reagent_containers/mender) && !istype(over_object, /obj/item/tank/jetpack/backtank) && !istype(over_object, /obj/machinery/bathtub))
+			return ..()
+		if (usr.stat || usr.getStatusDuration("weakened") || BOUNDS_DIST(usr, src) > 0 || BOUNDS_DIST(usr, over_object) > 0)  //why has this bug been in since i joined goonstation and nobody even looked here yet wtf -ZeWaka
+			boutput(usr, "<span class='alert'>That's too far!</span>")
+			return
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:262]: see other comment about your `mouse_drop` overrides

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 9]
```diff
@@ -114,117 +147,178 @@
...
-			if(src.reagents.total_volume > 5)
-				reagents.reaction(src.myuser, TOUCH)
-				src.reagents.trans_to(src.myuser,5,1)
-			else
-				src.reagents.clear_reagents()
-				on_reagent_change()
-
-		return
+			if(src.reagents.total_volume)
+				src.reagents.reaction(src.occupant, TOUCH)
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:281]: Again, should probably have a reaction volume, and not just apply all 500u. Also should use the returned list from `reaction()` to consume reagents with an on-touch effect like styptic/sulf/acid/etc

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 10]
```diff
@@ -95,16 +95,49 @@
...
+			animate_wiggle_then_reset(src, 1, 3)
+			src.reagents.trans_to(src.last_turf, 5)
+		else
+			..()
+
+	// dirty copy/paste hack from breath & fluid_core
+	handle_internal_lifeform(mob/lifeform_inside_me, breath_request)
+		var/mob/M = lifeform_inside_me
+		if (M.lying && M.get_oxygen_deprivation() > 40 && src.reagents.total_volume > suffocation_volume && breath_request > 0 )
+			src.reagents.reaction(M, INGEST)
```
[the remark this answers is not in the document]
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: Re-worked this to be emulating turf fluid puddles more accurately.

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 11]
```diff
@@ -95,16 +95,49 @@
...
 	icon_state = "bathtub"
 	flags = OPENCONTAINER
-	var/static/image/fluid_icon
-	var/static/image/fluid_icon_feet = image("icon" = 'icons/obj/stationobjs.dmi', "icon_state" = "fluid_bathtub_feet")
-	var/static/image/bath_edge = image("icon" = 'icons/obj/stationobjs.dmi', "icon_state" = "bath_edge", "layer" = FLOAT_LAYER)
-	var/mob/living/carbon/human/myuser = null
+	var/mob/living/carbon/human/occupant = null
 	var/default_reagent = "water"
+	var/on = FALSE
+	var/suffocation_volume = 200 // tracking fluid_core's drowning depth_level vOv
```
[the remark this answers is not in the document]
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: Changed, thank you!

File: code/WorkInProgress/DrMelonsStuff.dm (hunk 9 above)
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: Reworked this to match how fluids handle touch reactions to match behavior

**Comment by pgmzeta** [comment]
Thank you for the questions on INGEST/TOUCH volume - making bathtubs work like turfs was a clearer direction to go in.

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 12]
```diff
@@ -196,17 +203,7 @@
...
-		src.occupant.layer = initial(occupant.layer)
-		src.occupant.set_loc(get_turf(src))
-		src.vis_contents -= src.occupant
-		src.occupant = null
-		src.UpdateOverlays(null, "fluid_overlay")
-		src.UpdateOverlays(null, "bath_edge")
-		src.on_reagent_change()
-		if (src.reagents.total_volume)
-			playsound(src.loc, "sound/misc/splash_1.ogg", 70, 3)
+		Exited(src.occupant, user.loc)
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:212]: You shouldn't be explicitly calling `Exited()` -- set_loc will do it for you

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 13]
```diff
@@ -108,9 +108,9 @@
 			desc += " But, it's faintly fizzing?"
 
 	disposing()
-		if(src.occupant)
-			src.occupant.set_loc(get_turf(src))
-			src.occupant = null
+		if (src.occupant) Exited(src.occupant, get_turf(src))
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:111]: You shouldn't be explicitly calling Exited() -- set_loc will do it for you

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 14]
```diff
@@ -299,6 +284,24 @@
 					src.occupant.update_clothing()
 					src.occupant.update_body()
 
+	Exited(atom/movable/Obj, loc)
+		..()
+		if (ismob(Obj))
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:302]: should check that it's the occupant that's exiting, not just any mob

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 15]
```diff
@@ -299,6 +284,24 @@
...
+			src.visible_message("<span class='notice'>[src.occupant] gets out of the bath.</span>", "<span class='notice'>You get out of the bath.</span>")
+			if (src.reagents.total_volume)
+				playsound(src.loc, "sound/misc/splash_1.ogg", 70, 3)
+			var/mob/M = Obj
+			M.pixel_y = 0
+			M.layer = initial(src.occupant.layer)
+			M.loc = loc
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:309]: do not explicitly set `loc`

Comment (Tarmunora) [reply code/WorkInProgress/DrMelonsStuff.dm:309]: (or call set_loc on the occupant, because `Exited()` is called when they leave)

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 16]
```diff
@@ -248,20 +245,8 @@
...
-			return
-		if (!istype(over_object, /obj/item/reagent_containers/glass) && !istype(over_object, /obj/item/reagent_containers/food/drinks) && !istype(over_object, /obj/reagent_dispensers) && !istype(over_object, /obj/item/spraybottle) && !istype(over_object, /obj/machinery/plantpot) && !istype(over_object, /obj/mopbucket) && !istype(over_object, /obj/item/reagent_containers/mender) && !istype(over_object, /obj/item/tank/jetpack/backtank) && !istype(over_object, /obj/machinery/bathtub))
-			return ..()
-		if (usr.stat || usr.getStatusDuration("weakened") || BOUNDS_DIST(usr, src) > 0 || BOUNDS_DIST(usr, over_object) > 0)  //why has this bug been in since i joined goonstation and nobody even looked here yet wtf -ZeWaka
-			boutput(usr, "<span class='alert'>That's too far!</span>")
-			return
-
-		src.transfer_all_reagents(over_object, usr)
+	relaymove(mob/user)
+		Exited(user, src.loc)
```
Comment (Tarmunora) [inline code/WorkInProgress/DrMelonsStuff.dm:252]: Again, don't call `Exited()` explicitly.

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/mob/living/life/breath.dm [hunk 17]
```diff
@@ -102,12 +102,10 @@
...
 			return
-
-		// kludge: fake fluid to trigger underwater suffocation check
 		else if (istype(owner.loc, /obj/machinery/bathtub) && owner.lying)
 			var/obj/machinery/bathtub/B = owner.loc
 			if (B.reagents.total_volume > B.suffocation_volume)
-				var/obj/fluid/F = new
+				var/obj/fluid/F = new // used for underwater breathing check
 				F.reagents = owner.loc.reagents
 				underwater = F
```
Comment (Tarmunora) [inline code/mob/living/life/breath.dm:110]: is any of this needed, or could it all be handled in handle_internal_lifeform?

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/mob/living/life/breath.dm (hunk 17 above)
Comment (pgmzeta) [reply code/mob/living/life/breath.dm]: The problem I am running into with trying to remove this is the interaction with underwater checks in [breath.dm:131](https://github.com/goonstation/goonstation/blob/fc36eb199cb420764be01e82c82da5f2182eac68/code/mob/living/life/breath.dm#L131=). That code handles returning early when someone is an aquatic mutrantrace, has the 'aquabreath' status (no datum?), or messaging the user if they're drowning.

I'm not sure what the best options are here; rework object handling for `breath.dm` (in a separate PR?), shelve this feature, or go forward with the hacky solution.

File: code/WorkInProgress/DrMelonsStuff.dm (hunk 12 above)
Comment (pgmzeta) [reply code/WorkInProgress/DrMelonsStuff.dm]: thank you, I misunderstood how Exited was used; these should all be fixed to use `set_loc`

**Comment by Tarmunora** [comment]
wgcw
